// testarmada-magellan-saucelabs-executor/lib/logger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _yargs = require("yargs");

var _util = require("util");

var _util2 = _interopRequireDefault(_util);

var _cliColor = require("cli-color");

var _cliColor2 = _interopRequireDefault(_cliColor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _debug = _yargs.argv.debug;
var PREFIX = "Saucelabs Executor";

exports.default = {
  output: console,

  debug: function debug(msg) {
    if (_debug) {
      var deb = _cliColor2.default.blueBright("[DEBUG]");
      this.output.log(_util2.default.format("%s [%s] %s", deb, PREFIX, msg));
    }
  },
  log: function log(msg) {
    var info = _cliColor2.default.greenBright("[INFO]");
    this.output.log(_util2.default.format("%s [%s] %s", info, PREFIX, msg));
  },
  warn: function warn(msg) {
    var warn = _cliColor2.default.yellowBright("[WARN]");
    this.output.warn(_util2.default.format("%s [%s] %s", warn, PREFIX, msg));
  },
  err: function err(msg) {
    var err = _cliColor2.default.redBright("[ERROR]");
    this.output.error(_util2.default.format("%s [%s] %s", err, PREFIX, msg));
  },
  loghelp: function loghelp(msg) {
    this.output.log(msg);
  },
  stringifyLog: function stringifyLog(msg) {
    var info = _cliColor2.default.greenBright("[INFO]");
    return _util2.default.format("%s [%s] %s", info, PREFIX, msg);
  },
  stringifyWarn: function stringifyWarn(msg) {
    var warn = _cliColor2.default.yellowBright("[WARN]");
    return _util2.default.format("%s [%s] %s", warn, PREFIX, msg);
  },
  stringifyErr: function stringifyErr(msg) {
    var err = _cliColor2.default.redBright("[ERROR]");
    return _util2.default.format("%s [%s] %s", err, PREFIX, msg);
  }
};
