// testarmada-magellan-saucelabs-executor/lib/settings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _yargs = require("yargs");

var _yargs2 = _interopRequireDefault(_yargs);

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var debug = _yargs2.default.argv.debug;
var TEMP_DIR = _path2.default.resolve(_yargs2.default.argv.temp_dir || "./temp");

/*eslint-disable no-magic-numbers*/
var config = {
  tunnel: {
    // required:
    username: null,
    accessKey: null,
    connectVersion: null,

    // optional:
    tunnelIdentifier: null,
    fastFailRegexps: null
  },
  sharedSauceParentAccount: null,
  useTunnels: false,

  sauceOutboundProxy: null,

  app: null,
  appCapabilitiesConfig: null,

  locksServerLocation: null,
  locksOutageTimeout: 1000 * 60 * 5,
  locksPollingInterval: 5000,
  locksRequestTimeout: 5000
};

exports.default = {
  debug: debug,
  tempDir: TEMP_DIR,
  config: config,

  MAX_CONNECT_RETRIES: process.env.SAUCE_CONNECT_NUM_RETRIES || 10,
  BASE_SELENIUM_PORT_OFFSET: 56000,
  BAILED: false
};
