// testarmada-nightwatch-extra/lib/commands/mobile/getMobileElValue.js
"use strict";

var _util = require("util");

var _util2 = _interopRequireDefault(_util);

var _baseMobileCommand = require("../../base-mobile-command");

var _baseMobileCommand2 = _interopRequireDefault(_baseMobileCommand);

var _settings = require("../../settings");

var _settings2 = _interopRequireDefault(_settings);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var GetMobileElValue = function GetMobileElValue() {
  var nightwatch = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

  _baseMobileCommand2.default.call(this, nightwatch);
  this.cmd = "getmobileels";
};

_util2.default.inherits(GetMobileElValue, _baseMobileCommand2.default);

GetMobileElValue.prototype.do = function (value) {
  var self = this;

  this.client.api.elementIdValue(value.ELEMENT, function (result) {
    if (result.status === 0) {
      self.pass({ actual: result.value });
    } else {
      self.fail({
        code: _settings2.default.FAILURE_REASONS.BUILTIN_ELEMENT_NOT_OPERABLE,
        message: result.error
      });
    }
  });
};

GetMobileElValue.prototype.command = function (using, selector, cb) {
  this.selector = selector;
  this.using = using;
  this.cb = cb;

  this.successMessage = "Selector '" + this.using + ":" + this.selector + "' " + "was visible after %d milliseconds.";
  this.failureMessage = "Selector '" + this.using + ":" + this.selector + "' " + "was not visible after %d milliseconds.";

  this.startTime = new Date().getTime();

  // Track how many times selector is successfully checked by /element protocol
  this.seenCount = 0;
  this.checkConditions();

  return this;
};

module.exports = GetMobileElValue;
